### DISCUSSION TITLE
incompatible pointer and type mismatched when importing FMU from OM

### DISCUSSION THREAD

**Discussion by Arinomo** [OpenModelica/OpenModelica#11436] [Q&A] [answered]
I was trying to compile a rt-application using dSPACE with FMU exported from OpenModelica v1.22.0-dev-39-g23fdc84c46. It worked with the prior version (I think 1.20), but today I got following error about incompatible pointer type and type mismatched:

![image](https://github.com/OpenModelica/OpenModelica/assets/52284681/f5fd9ad1-c270-4656-b297-02fc223236b7)

Any idea why or what trigger the error?

I exported the fmu on Win10 PC. dSpace took codes from source directory and used it to compile the rt-application.

Simple model that I exported:
 ```.mo
model LottkaVoltera
 parameter Real alpha=0.1 "Reproduction rate of prey";
  parameter Real beta=0.02 "Mortality rate of predator per prey";
  parameter Real gamma=0.4 "Mortality rate of predator";
  parameter Real delta=0.02 "Reproduction rate of predator per prey";
  parameter Real x0=10 "Initial prey population";
  parameter Real y0=10 "Initial predator population";
  Real x(start=x0) "Prey population";
  Real y(start=y0) "Predator population";
initial equation
  x = x0;
  y = y0;
equation
  der(x) = x*(alpha-beta*y);
  der(y) = y*(delta*x-gamma);

end LottkaVoltera;
 ```

**Comment by AnHeuermann** [comment]
`comp->threadData` has type `threadData_t*` and that should be the type of the second input to `handleTimersFMI`

```C
int handleTimersFMI(DATA* data, threadData_t *threadData, double currentTime, int *nextTimerDefined, double *nextTimerActivationTime)
```

Maybe the declaration of `handleTimersFMI` is missing? It should be in `/sources/simulation/solver/synchronous.h`.


Can you try again with a nightly version of OpenModelica?

**Reply by casella** [reply]
I would suggest you try both with the latest nightly and with 1.22.0-dev.beta.6.

If both work, we're done. If only the second does, we need to port some fix to 1.22.0. If none works, then we can try to figure out why this happens.

**Reply by Arinomo** [reply]
@casella ok, i will export the fmu from both versions and see how it goes.

@AnHeuermann:
If I understand the error message correctly, it was complaining about argument 4 of `handlerTimersFMI`, so its either `int *nextTimerDefined` or `double *nextTimerActivationTime`.

Yes the declaration is in the `synchronous.h` with return type `int`, hence the second error message that complaints about getting `fmi2Boolean` type as a return.

Edit:
After looking at it this morning, I think the type mismatched is referring to the type of argument 4 of `handlersTimersFMI`. In `synchronous.h` argument 4 is defined as `int *nextTimerDefined`.

In [fmu2_model_interface.c.inc Line 315](https://github.com/OpenModelica/OpenModelica/blob/a073984273e75e4c39fe569c3b97e00945f38d90/OMCompiler/SimulationRuntime/fmi/export/openmodelica/fmu2_model_interface.c.inc#L315C3-L315C3) `&nextTimerDefined` of type `fmi2Boolean` was passed to `handlerTimerFMI`. Maybe that what triggered the error message?

**Reply by AnHeuermann** [reply, marked as answer]
Oh yes. I got confused from the image where the `^` is at the second input.

`fmi2Boolean` isn't always compatible to `int`. Can you change it to `modelica_boolean` in synchronous.c / synchronous.h and see if that solves the issue?

```diff
-int handleTimersFMI(DATA* data, threadData_t *threadData, double currentTime, int *nextTimerDefined, double *nextTimerActivationTime)
+int handleTimersFMI(DATA* data, threadData_t *threadData, double currentTime, modelica_boolean *nextTimerDefined, double *nextTimerActivationTime)
```

**Reply by Arinomo** [reply]
@AnHeuermann you are right, changing the type to `modelica_boolean` solve the issue. I created a [PR#11469](https://github.com/OpenModelica/OpenModelica/pull/11469) for this.
